### DISCUSSION TITLE
Parsing error reasons

### DISCUSSION THREAD

**Discussion by JasterV** [stalwartlabs/mail-parser#57] [Q&A] [answered]
Hello there! First of all, congratulations for the great work. This library is amazing.
I work for a company where we have to deal with very old email clients sending malformed emails that rarely conform to an RFC and this library just parses them like a charm!!

I have one question tho. The `Message::parse` function just returns an `Option` type, why? Aren't there many reasons on why the parsing could fail? Is it really just the missing headers reason the only one? Hard to believe since we all know dealing with emails and all these different RFCs is quite a nightmare, I expected the `parse` function to return me an error with a more detailed message on what happened.

Hope I can get an answer, thank you!

**Comment by mdecimus** [comment, marked as answer]
Hi,

`None` is returned when not a single valid header is found in the message. In all other cases, if at least one header is found (including non-RFC headers), then `Some()` will be returned.

For example:

```txt
Hello: World
```

Will return a parsed message with a single header `Hello` with value `World.`. However, when parsing:

```txt
Hello World
```

A `None` will be returned as there is not a single header in the input.

**Reply by JasterV** [reply]
Cool!

What about the body and all the other message parts? Isn't there any case on why you could return an error due to that?
Just curious about it since I'm now implementing all the error handling on my application and this library plays a very important role on it
Amazing work!!

**Reply by mdecimus** [reply]
If the MIME message is broken (i.e. no boundaries are found), then the message is parsed as plain text.

If there is a problem decoding a message part then the raw text part is provided instead and [is_encoding_problem](https://docs.rs/mail-parser/latest/mail_parser/struct.MessagePart.html#structfield.is_encoding_problem) is set to `true`. Oops, I just found out this is not documented anywhere!

**Reply by JasterV** [reply]
Oh interesting! Thank you :) 

How should I handle this then? Lets say I use the `parse` function and I get a `Message`
Then I want to get all the attachments from this message (Thats kinda my use case), which are message parts.
I use the `attachments` iterator to iterate over the message parts and get the content + some other data like name and content type
Is it possible that I get a `MessagePart` type from this iterator that is not well parsed? Should then I check the `is_encoding_problem` boolean and treat that as an error?

**Reply by mdecimus** [reply]
Yes, you can always iterate the message parts even if there were errors decoding. Then, if while iterating you find an `is_encoding_problem` set to true you can treat that as an error.

**Reply by JasterV** [reply]
amazing thanks!!
Would be nice to have a way to get more context about the error tho. Lets say I have a bunch of emails that triggered an error, if its a big email I will probably spend some hours trying to understand what is wrong with it and maybe its a very specific character that made it fail

But anyway, this already answers my question :) 
Thank you!

**Reply by JasterV** [reply]
@mdecimus 

As an idea, instead of using a boolean I would refactor the `MessagePart` type to be an enum with 2 variants.
1 variant would represent a correctly parsed `MessagePart` and the other would represent a part that failed to parse, still exposing the raw content + whatever data you are able to expose + some context on why the parsing failed
This way you are enforcing everyone to always handle the case where the parsing of a message part failed :D

**Reply by mdecimus** [reply]
Makes sense, however the API is trying to follow the JMAP specification in which the contents are provided even in case of failure and a boolean `is_encoding_problem` variable is set to `true`.

**Reply by JasterV** [reply]
I see! The solution I provided would still provide the contents even on failure tho!
